internal/engine/upper.go

```diff
@@ -224,6 +224,94 @@ func handleBuildStarted(ctx context.Context, state *store.EngineState, action bu
 	removeFromTriggerQueue(state, mn)
 }
 
+func updateBuildStatus(status *store.BuildStatus, br model.BuildRecord, result store.BuildResult, isBuildSuccess bool) {
+	// Record the last result
+	status.LastResult = result
+
+	// Remove pending file changes that were consumed by this build.
+	for file, modTime := range status.PendingFileChanges {
+		if store.BeforeOrEqual(modTime, br.StartTime) {
+			delete(status.PendingFileChanges, file)
+		}
+	}
+
+	// Remove pending dependency changes that were consumed by this build.
+	for id, modTime := range status.PendingDependencyChanges {
+		if store.BeforeOrEqual(modTime, br.StartTime) {
+			delete(status.PendingDependencyChanges, id)
+		}
+	}
+
+	// If the whole build succeeded, mark all of the images in the
+	// build as successful.
+	//
+	// This has a minor bug: if one of the images in a set failed, but another one
+	// succeeded, this doesn't mark any of them as successful. We should fix this,
+	// though personally, I think LastSuccessfulResult has problematic semantics
+	// to begin with, and it might make sense to replace it entirely.
+	if isBuildSuccess {
+		status.LastSuccessfulResult = result
+	}
+}
+
+// When a Manifest build finishes, update the BuildStatus for all applicable
+// targets in the engine state.
+func handleBuildResults(engineState *store.EngineState,
+	mt *store.ManifestTarget, br model.BuildRecord, results store.BuildResultSet) {
+	isBuildSuccess := br.Error == nil
+
+	// Update all the build statuses for the current manifest.
+	for _, target := range mt.Manifest.TargetSpecs() {
+		status := mt.State.MutableBuildStatus(target.ID())
+		updateBuildStatus(status, br, results[target.ID()], isBuildSuccess)
+	}
+
+	// Update build statuses for duplicated image targets in other manifests.
+	// This ensures that those images targets aren't rebuilt.
+	engineState.ForEachMutableBuildStatusInResults(results, func(id model.TargetID, result store.BuildResult, mn model.ManifestName, status *store.BuildStatus) {
+		if mt.Manifest.Name == mn {
+			return
+		}
+
+		_, isImageResult := result.(store.ImageBuildResult)
+		if !isImageResult {
+			return
+		}
+
+		updateBuildStatus(status, br, result, isBuildSuccess)
+	})
+
+	// Suppose we built manifestA, which contains imageA depending on imageCommon.
+	//
+	// We also have manifestB, which contains imageB depending on imageCommon.
+	//
+	// We need to mark imageB as dirty, because it was not built in the manifestA
+	// build but its dependency was built.
```

Comment:
(idk if this comment is quite right but i DO think it's useful to mention the deploy-targets-depending-on-image-targets case--when this comment only talks about inter-image dependencies, it seems kinda special-case-y, but really this bit of code is much more generic than that!)
```suggestion
	// build but its dependency was built.
	
	// Note that this logic also applies to deploy targets depending on image targets. If we built manifestA, which contains imageX and deploy target k8sA, and manifestB contains imageX and deploy target k8sB, we need to mark target k8sB as dirty so that Tilt actually deploys the changes to imageX.
```

Reply:
done